driver/src/orderbook/streamed/balance.rs

```diff
@@ -2,63 +2,104 @@ use super::*;
 use error::Error;
 use serde::{Deserialize, Serialize};
 
+/// The balance of a token for a user.
 #[derive(Clone, Copy, Debug, Default, Deserialize, Serialize)]
 pub struct Balance {
     balance: U256,
-    pub pending_deposit: Option<Flux>,
-    pub pending_withdraw: Option<Flux>,
+    deposit: Flux,
+    withdraw: Flux,
 }
 
 impl Balance {
-    pub fn deposit(&mut self, new_deposit: Flux, current_batch_id: BatchId) {
-        self.update_deposit_balance(current_batch_id);
-        match self.pending_deposit.as_mut() {
-            Some(deposit) => {
-                deposit.amount += new_deposit.amount;
-                deposit.batch_id = new_deposit.batch_id
-            }
-            None => self.pending_deposit = Some(new_deposit),
-        }
+    pub fn deposit(
+        &mut self,
+        amount: U256,
+        deposit_batch_id: BatchId,
+        current_batch_id: BatchId,
+    ) -> Result<(), Error> {
+        self.apply_existing_deposit(current_batch_id)?;
+        // Works like in the smart contract: If there is an existing deposit we override the
+        // batch id and add to the amount. If there is no existing deposit then amount is already 0.
+        let deposit_amount_including_this_deposit = self
+            .deposit
+            .amount
+            .checked_add(amount)
+            .ok_or(Error::MathOverflow)?;
+        self.deposit.batch_id = deposit_batch_id;
+        self.deposit.amount = deposit_amount_including_this_deposit;
+        Ok(())
+    }
+
+    pub fn withdraw_request(&mut self, amount: U256, batch_id: BatchId) {
+        self.withdraw.batch_id = batch_id;
+        self.withdraw.amount = amount;
```

Comment:
Once a balance withdrawal is requested, is the balance still available to the exchange and the orderbook for trading or is it "useless"?

Reply:
Do you mean once the effective balance is 0 because of the valid withdraw request is >= balance? The smart contract uses `getBalance` so it will also see 0 and can't deduct any more.

Reply:
Sorry, I mean if user A has a balance of T OWL and makes a withdrawal request of T in batch X (that matures in batch X+1), then is user A's balance in batch X+1 T or 0?

The question is should withdrawal requests be updating the balance or should withdrawals be? In the UI it looks like a withdrawal request makes it so your balance is no longer in the exchange.

Reply:
But at the same time I know that there was a withdrawal denial attack where if a balance gets touched while there is a withdrawal request, then it changes its maturing batch to the one after (or something like that - I don't remember the exact details).

Reply:
In this code it is supposed to work like in the smart contract, so that the solver sees the balance that the smart contract also sees (like in the paginated orderbook). This is `get_balance`. In your example in batch X+1 `get_balance` returns 0 because the withdrawal request is valid.
If I understand you correctly you're asking whether we can simplify the code by deducting the withdrawal request balance immediately and ignoring withdraw events.
We cannot. Example:

1. it is batch X
2. user U has balance B0
3. user U requests withdrawal RW > B0
4. user U gains balance B1 from a trade
5. it is batch X + 1
6. user withdraws RW

The correct final balance here is `(B0 + B1) - RW`.

But if we ignored withdraws and applied withdraw requests immediately then we would set the balance to 0 in step 3, increase it to B1 from 0 in step 4, ignore the withdraw in step 6 and en dup with balance B1.

Reply:
> The correct final balance here is (B0 + B1) - RW.

But this is regardless of step 6 right? so the balance is `(B0 + B1) - RW` even without "user withdraws RW"?

Reply:
Yes. I included that step to show that the user withdrew RW and not just B0. The point is that we cannot remove the withdraw request until we get the corresponding withdraw to see how much was actually withdrawn.

Reply:
So, lets say the balance gets updated with all the events in a batch at once, such that the `Balances` state no longer needs to expect any other event that will come in for a given batch. Can't the new balance simply be computed by as `new_balance = balance + sum(deposit amounts + trade buy amounts) - sum(withdrawal amounts + trade sell amounts)` and the pending withdrawal be `pending_withdrawal = if withdrawn since last request { 0 } else { last(withdrawal request) }` making `available balace () => new_balance.saturating_sub(pending_withdrawal)`

But I see your point that withdrawal events are also needed since they clear the pending withdrawals.

I still think having events come in a collection is easier to manage cognitively and is in general more powerful since the balances state will have all relevant information upfront.

Reply:
So my point is that if state gets updated with event collections instead of singe events, then they can express their updates in simpler terms since have more information (the complete event list for a specific batch).

Note that this should still work as a continuous orderbook since the event collection will at that point just be the known events.

Another thing worth mentioning is that we can use `FutureExt::immediate` to get "ready" events so that the event stream can synchronously, without blocking, produce all events that came in a single `eth_getFilterChanges` call - so all the events within a block so you don't have to worry about partially applied blocks.

Reply:
I think this would work (but it's hard to tell without trying it) but the simplification comes from our not checking some errors that we are checking in the current version. If we don't check that a withdraw matches a withdraw request or doesn't withdraw too much then code is simpler but also less accurate. It would accept some order of events that could not really have occurred. I don't know how important this is but I went for accuracy when possible which is why there are so many error variants.
The one thing the batching removes is keeping track of deposits because indeed we can apply them all together. Pending withdraws we still need to keep track of because they can be for a future batch.
One problem I see now with batching is that we won't know when a batch is over. We will know if we get a new event that is in the next batch but we won't know know until then.

Reply:
> I went for accuracy when possible 

Good point, but I still think it would be easier to be accurate with an event collection as there is more information available to the state.

> One problem I see now with batching is that we won't know when a batch is over. We will know if we get a new event that is in the next batch but we won't know know until then.

We do based on block number. We can setup a block stream and then whenever we get a new block decide whether or not we can commit a snapshot (something like 5 blocks into the next batch would be ultra safe).

Reply:
> Pending withdraws we still need to keep track of because they can be for a future batch.

Correct, but you can flatten the collection of pending withdrawals since you only care about the last one.

Reply:
I do like your approach of checking every event as it comes in - I think its the best way to guarantee that the events are being applied correctly since the state would become invalid at some point if that weren't the case.

That being said, I think if we take the simpler less correct approach we can get up and running a bit quicker, and then still validate each event within the context of a single `apply_events` with some future added logic.

Reply:
> but the simplification comes from our not checking some errors that we are checking in the current version

I agree, the end goal should be to check everything so we can be sure our state is coherent, but I would suggest having that come later.